database/manager.go

```diff
@@ -0,0 +1,267 @@
+	for _, key := range keys {
+		keyManager := newKeyManager(logger, key)
+		go keyManager.run()
+		keyManagers[key] = keyManager
+	}
+	return &DatabaseManager{
+		logger:      logger,
+		db:          db,
+		interval:    interval,
+		keyManagers: keyManagers,
+	}
+}
+
+// Run writes all committed keys to the database on the configured timer.
+// This function should only be called once.
+func (dm *DatabaseManager) Run() {
+	for range time.Tick(dm.interval) {
+		for id, km := range dm.keyManagers {
+			// Ensure we're not writing the default value
+			km.lock.RLock()
+			currMaxCommittedHeight := km.maxCommittedHeight
+			km.lock.RUnlock()
+			if currMaxCommittedHeight == 0 {
+				continue
+			}
+			storedHeight, err := getLatestProcessedBlockHeight(dm.db, id)
+			if err != nil && !IsKeyNotFoundError(err) {
+				dm.logger.Error(
+					"Failed to get latest processed block height",
+					zap.Error(err),
+					zap.String("relayerID", id.ID.String()),
+				)
+				continue
+			}
+			if storedHeight >= currMaxCommittedHeight {
+				continue
+			}
+			dm.logger.Debug(
+				"Writing height",
+				zap.Uint64("height", currMaxCommittedHeight),
+				zap.String("relayerID", id.ID.String()),
+			)
+			err = dm.db.Put(id.ID, LatestProcessedBlockKey, []byte(strconv.FormatUint(currMaxCommittedHeight, 10)))
+			if err != nil {
+				dm.logger.Error(
+					"Failed to write latest processed block height",
+					zap.Error(err),
+					zap.String("relayerID", id.ID.String()),
+				)
+				continue
+			}
+		}
+	}
+}
+
+func (dm *DatabaseManager) Get(id RelayerID, key DataKey) ([]byte, error) {
+	return dm.db.Get(id.ID, key)
+}
+
+func (dm *DatabaseManager) PrepareHeight(id RelayerID, height uint64, totalMessages uint64) error {
+	km, ok := dm.keyManagers[id]
+	if !ok {
+		dm.logger.Error("Key manager not found", zap.String("relayerID", id.ID.String()))
+		return fmt.Errorf("key manager not found")
+	}
+	km.prepareHeight(height, totalMessages)
+	return nil
+}
+
+func (dm *DatabaseManager) Finished(id RelayerID, height uint64) {
+	km, ok := dm.keyManagers[id]
+	if !ok {
+		dm.logger.Error("Key manager not found", zap.String("relayerID", id.ID.String()))
+		return
+	}
+	km.finished <- height
+}
+
+// intHeap adapted from https://pkg.go.dev/container/heap#example-package-IntHeap
+type intHeap []uint64
+
+func (h intHeap) Len() int           { return len(h) }
+func (h intHeap) Less(i, j int) bool { return h[i] < h[j] }
+func (h intHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
+func (h intHeap) Peek() uint64       { return h[0] }
+
+func (h *intHeap) Push(x any) {
+	// Push and Pop use pointer receivers because they modify the slice's length,
+	// not just its contents.
+	*h = append(*h, x.(uint64))
+}
+
+func (h *intHeap) Pop() any {
+	old := *h
+	n := len(old)
+	x := old[n-1]
+	*h = old[0 : n-1]
+	return x
+}
+
+//
+// keyManager commits keys to be written to the database in a thread safe manner.
+//
+
+type keyManager struct {
```

Comment:
Small suggestion that may be missing some context, but would `BlockCheckpointManager` be a more descriptive name rather than `keyManager`? i.e. each `ApplicationRelayer` instance would have a `BlockCheckpointManager` that it uses to determine the correct value to write to the database.